src/foundry/client/canvas/placeables/tokens/ruler.d.mts

```diff
@@ -1,13 +1,211 @@
-import type { Identity } from "#utils";
+import type { DeepReadonly, Identity } from "#utils";
 import type { BaseTokenRuler } from "./_module.d.mts";
 
 declare class TokenRuler extends BaseTokenRuler {
   #TokenRuler: true;
+
+  /**
+   * A handlebars template used to render each waypoint label.
+   */
+  static WAYPOINT_LABEL_TEMPLATE: string;
+
+  protected override _onVisibleChange(): void;
+
+  /**
+   * Configure the properties of the outline.
+   * Called in {@link TokenRuler#draw}.
+   * @returns The thickness in pixels and the color
+   */
+  protected _configureOutline(): TokenRuler.Outline;
+
+  /**
+   * Configure the properties of the dash line.
+   * Called in {@link TokenRuler#draw}.
+   * @returns The dash in pixels, the gap in pixels, and the speed in pixels per second
+   */
+  protected _configureDashLine(): TokenRuler.DashLine;
+
+  draw(): Promise<void>;
+
+  clear(): void;
+
+  destroy(): void;
+
+  refresh(rulerData: TokenRuler.Data): void;
+
+  /**
+   * Get the context used to render a ruler waypoint label.
+   */
+  protected _getWaypointLabelContext(
+    waypoint: TokenRuler.Waypoint,
+    state: TokenRuler.State,
+  ): TokenRuler.WaypointContext | void;
+
+  /**
+   * Get the style of the waypoint at the given waypoint.
+   * @param waypoint - The waypoint
+   * @returns  The radius, color, and alpha of the waypoint. If the radius is 0, no waypoint marker is drawn.
+   */
+  protected _getWaypointStyle(waypoint: TokenRuler.Waypoint): TokenRuler.WaypointStyle;
+
+  /**
+   * Get the style of the segment from the previous to the given waypoint.
+   * @param waypoint - The waypoint
+   * @returns The line width, color, and alpha of the segment.  If the width is 0, no segment is drawn.
+   */
+  protected _getSegmentStyle(waypoint: TokenRuler.Waypoint): TokenRuler.SegmentStyle;
+
+  /**
+   * Get the style to be used to highlight the grid offset.
+   * @param waypoint - The waypoint
+   * @param offset - An occupied grid offset at the given waypoint that is to be highlighted
+   * @returns The color, alpha, texture, and texture matrix to be used to draw the grid space.
+   *          If the alpha is 0, the grid space is not highlighted.
+   */
+  protected _getGridHighlightStyle(
+    waypoint: TokenRuler.Waypoint,
+    offset: DeepReadonly<foundry.grid.BaseGrid.Offset3D>,
```

Comment:
So my policy on readonly is that we only apply it when it happens at runtime. This has only come up more recently as Foundry has begun to use it more often.

The reason for this policy goes as follows:
1. Readonly inputs are already accepted (you can assign readonly inputs to mutable ones due to unsoundness).
2. For a readonly output _someone_ is going to want to mutate it or arguably more important they'll want to use it somewhere else where Foundry hasn't updated to allow readonly.